crates/ika-core/src/dwallet_mpc/mpc_manager.rs

```diff
@@ -635,27 +317,142 @@ impl DWalletMPCManager {
+            self.decryption_key_reconfiguration_third_round_delay,
         );
+
         info!(
-            // todo(zeev): add metadata.
+            party_id=self.party_id,
+            authority=?self.validator_name,
+            with_mpc_event_data,
+            ?session_identifier,
             last_session_to_complete_in_current_epoch=?self.last_session_to_complete_in_current_epoch,
-            "Adding MPC session to active sessions",
+            "Adding a new MPC session to the active sessions map",
         );
+
         self.mpc_sessions.insert(*session_identifier, new_session);
     }
 
+    /// Spawns all ready MPC cryptographic computations on separate threads using Rayon.
+    /// If no local CPUs are available, computations will execute as CPUs are freed.
+    ///
+    /// A session must have its `mpc_event_data` set in order to be advanced.
+    ///
+    /// System sessions are always advanced if a CPU is free, user sessions are only advanced
+    /// if they come before the last session to complete in the current epoch (at the current time).
+    ///
+    /// System sessions are always advanced before any user session,
+    /// and both system and user sessions are ordered internally by their sequence numbers.
+    ///
+    /// The messages to advance with are built on the spot, assuming they satisfy required conditions.
+    /// They are put on a `ComputationRequest` and forwarded to the `orchestrator` for execution.
+    pub(crate) async fn perform_cryptographic_computation(
+        &mut self,
+    ) -> HashMap<
+        ComputationId,
+        DwalletMPCResult<
+            mpc::AsynchronousRoundResult<
+                MPCMessage,
+                MPCPrivateOutput,
+                SerializedWrappedMPCPublicOutput,
+            >,
+        >,
+    > {
+        let computation_requests: Vec<_> = self
+            .mpc_sessions
+            .iter()
+            .filter(|(_, session)| {
+                if let Some(mpc_event_data) = &session.mpc_event_data {
+                    // Always advance system sessions, and only advance user session
+                    // if they come before the last session to complete in the current epoch (at the current time).
+                    match mpc_event_data.session_type {
+                        SessionType::User => {
+                            mpc_event_data.session_sequence_number
+                                <= self.last_session_to_complete_in_current_epoch
+                        }
+                        SessionType::System => true,
+                    }
+                } else {
+                    // Cannot advance sessions without MPC event data
+                    false
+                }
+            })
+            .sorted_by(|(_, session), (_, other_session)| {
+                // Safe to `unwrap`, we filtered sessions without `mpc_event_data`.
+                let session_mpc_event_data = session.mpc_event_data.as_ref().unwrap();
```

Comment:
I will be cleaner to add mpc_event_data to the tuple so then you don't have to keep unwrap it

Reply:
As part of filter (maybe should be filter_map instead)